File: frontend/src/api/queue.ts
import type { QueueEntry } from '../types'
import { api } from './client'

export const queueApi = {
  list: () => api.get<QueueEntry[]>('/api/queue'),
  add: (postIds: string[], priority = 0) =>
    api.post<{ added: number }>('/api/queue', { post_ids: postIds, priority }),
  remove: (postId: string) => api.delete<{ removed: boolean }>(`/api/queue/${postId}`),
  updatePriority: (postId: string, priority: number) =>
    api.patch<{ ok: boolean }>(`/api/queue/${postId}`, { priority }),
  reorder: (postIds: string[]) =>
    api.post<{ ok: boolean }>('/api/queue/reorder', { post_ids: postIds }),
  start: () => api.post<Record<string, unknown>>('/api/queue/start'),
  pause: () => api.post<{ ok: boolean }>('/api/queue/pause'),
  clear: () => api.post<{ ok: boolean }>('/api/queue/clear'),
}

File: web/routes/queue.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Request
from pydantic import BaseModel

from web.database import get_db
from web.models import QueueAddRequest
from web.services.queue_service import QueueService

router = APIRouter()


class PriorityUpdate(BaseModel):
    priority: int


class ReorderRequest(BaseModel):
    post_ids: list[str]


def _queue_service(request: Request, db=Depends(get_db)) -> QueueService:
    return QueueService(db, request.app.state.obs_client)


@router.get("/")
async def get_queue(service: QueueService = Depends(_queue_service)):
    return await service.get_queue()


@router.post("/")
async def add_to_queue(body: QueueAddRequest, service: QueueService = Depends(_queue_service)):
    count = await service.add_to_queue(body.post_ids, body.priority)
    return {"added": count}


@router.delete("/{post_id}")
async def remove_from_queue(post_id: str, service: QueueService = Depends(_queue_service)):
    removed = await service.remove_from_queue(post_id)
    return {"removed": removed}


@router.patch("/{post_id}")
async def update_priority(
    post_id: str, body: PriorityUpdate, service: QueueService = Depends(_queue_service),
):
    await service.update_priority(post_id, body.priority)
    return {"ok": True}


@router.post("/reorder")
async def reorder(body: ReorderRequest, service: QueueService = Depends(_queue_service)):
    await service.reorder(body.post_ids)
    return {"ok": True}


@router.post("/start")
async def start_queue(service: QueueService = Depends(_queue_service)):
    return await service.start_queue()


@router.post("/pause")
async def pause_queue(service: QueueService = Depends(_queue_service)):
    await service.pause_queue()
    return {"ok": True}


@router.post("/clear")
async def clear_completed(service: QueueService = Depends(_queue_service)):
    await service.clear_completed()
    return {"ok": True}
